modules/pricer/src/main/java/com/opengamma/strata/pricer/impl/tree/TrinomialTree.java

```diff
@@ -0,0 +1,116 @@
+ */
+public class TrinomialTree {
+
+  /**
+   * Price an option under the specified trinomial lattice.
+   * <p>
+   * It is assumed that the volatility, interest rate and continuous dividend rate are constant 
+   * over the lifetime of the option. 
+   * 
+   * @param lattice  the lattice specification
+   * @param function  the option
+   * @param spot  the spot
+   * @param volatility  the volatility
+   * @param interestRate  the interest rate
+   * @param dividendRate  the dividend rate
+   * @return the option price
+   */
+  public double optionPrice(
+      LatticeSpecification lattice,
+      OptionFunction function,
+      double spot,
+      double volatility,
+      double interestRate,
+      double dividendRate) {
+
+    int nSteps = function.getNumberOfSteps();
+    double timeToExpiry = function.getTimeToExpiry();
+    double dt = timeToExpiry / (double) nSteps;
+    double discount = Math.exp(-interestRate * dt);
+    DoubleArray params = lattice.getParametersTrinomial(volatility, interestRate - dividendRate, dt);
+    double middleFactor = params.get(1);
+    double downFactor = params.get(2);
+    double upProbability = params.get(3);
+    double midProbability = params.get(4);
+    double downProbability = params.get(5);
+    ArgChecker.isTrue(upProbability > 0d, "upProbability should be greater than 0");
+    ArgChecker.isTrue(upProbability < 1d, "upProbability should be smaller than 1");
+    ArgChecker.isTrue(midProbability > 0d, "midProbability should be greater than 0");
+    ArgChecker.isTrue(midProbability < 1d, "midProbability should be smaller than 1");
+    ArgChecker.isTrue(downProbability > 0d, "downProbability should be greater than 0");
+    DoubleArray values = function.getPayoffAtExpiryTrinomial(spot, downFactor, middleFactor);
+    for (int i = nSteps - 1; i > -1; --i) {
+      values = function.getNextOptionValues(discount, upProbability, midProbability, downProbability, values, spot,
+          downFactor, middleFactor, i);
+    }
+    return values.get(0);
+  }
+
+  /**
+   * Price an option under the specified trinomial tree gird.
+   * 
+   * @param function  the option
+   * @param data  the trinomial tree data
+   * @return the option price
+   */
+  public double optionPrice(
+      OptionFunction function,
+      RecombiningTrinomialTreeData data) {
+
+    int nSteps = data.getNumberOfSteps();
+    ArgChecker.isTrue(nSteps == function.getNumberOfSteps(), "mismatch in number of steps");
+    DoubleArray values = function.getPayoffAtExpiryTrinomial(data.getStateValueAtLayer(nSteps));
+    for (int i = nSteps - 1; i > -1; --i) {
+      values = function.getNextOptionValues(
+          data.getDiscountFactorAtLayer(i), data.getProbabilityAtLayer(i), data.getStateValueAtLayer(i), values, i);
+    }
+    return values.get(0);
+  }
+
+  /**
+   * Compute option price and delta under the specified trinomial tree gird.
```

Comment:
Indicate it is the delta with respect to the spot price and the spot price is embedded in the tree data.